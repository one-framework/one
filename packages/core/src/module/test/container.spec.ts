import 'reflect-metadata';
import {
  Injectable,
  Module,
  OneContainer,
  UnknownProviderException,
} from '@one/core';

describe('OneContainer', () => {
  let container: OneContainer;

  @Module()
  class TestModule {}

  @Module()
  class TestModule2 {}

  @Injectable()
  class Nest {}

  beforeEach(() => {
    container = new OneContainer();
  });

  describe('getModules', () => {
    let getModuleValuesSpy: jest.SpyInstance;
    let getModuleSpy: jest.SpyInstance;

    beforeEach(() => {
      getModuleValuesSpy = jest.spyOn(container, 'getModuleValues');
      getModuleSpy = jest.spyOn(container, 'getModule');
    });

    afterEach(() => {
      getModuleValuesSpy.mockClear();
      getModuleSpy.mockClear();
    });

    it('should call getModuleValues if nil', async () => {
      await container.addModule(TestModule);
      const modules = (<any>container).getModulesFrom();
      const testModule = container.modules.values().next().value;

      expect(getModuleSpy).not.toHaveBeenCalled();
      expect(getModuleValuesSpy).toHaveBeenCalled();
      expect(modules[0]).toStrictEqual(testModule);
    });

    it('should call getModule if not nil', async () => {
      await container.addModule(TestModule);
      const modules = (<any>container).getModulesFrom(<any>TestModule);
      const testModule = container.modules.values().next().value;

      // Calls with [] ????
      // expect(getModuleValuesSpy).not.toHaveBeenCalled();
      expect(getModuleSpy).toReturnWith(testModule);
      expect(modules[0]).toStrictEqual(testModule);
    });
  });

  describe('isProviderBound', () => {
    it('should check if provider is bound in modules tree', async () => {
      await container.addModule(TestModule);
      const testModule = container.modules.values().next().value;

      await (<any>testModule).bindProvider(<any>Nest);

      expect(container.isProviderBound(Nest)).toBeTrue();
    });

    it('should check if provider is bound in specific module', async () => {
      await container.addModule(TestModule);
      await container.addModule(TestModule2);
      const testModule = container.modules.values().next().value;

      await (<any>testModule).bindProvider(<any>Nest);

      expect(container.isProviderBound(Nest, TestModule2)).toBeFalse();
      expect(container.isProviderBound(Nest, TestModule)).toBeTrue();
    });
  });

  describe('getProvider', () => {
    it('should get provider in nested modules tree', async () => {
      await container.addModule(TestModule);
      await container.addModule(TestModule2);
      const testModule2 = container.getModuleValues()[1];

      await (<any>testModule2).bindProvider(<any>Nest);

      await expect(container.getProvider(Nest, null)).toBeInstanceOf(Nest);
    });

    it('should lookup provider in scope when strict is true', async () => {
      await container.addModule(TestModule);
      const testModule = container.modules.values().next().value;

      await (<any>testModule).bindProvider(<any>Nest);

      expect(
        container.getProvider(Nest, TestModule, {
          strict: true,
        }),
      ).toBeInstanceOf(Nest);
    });

    it('should throw UnknownProviderException if not found', () => {
      const message = new UnknownProviderException(Nest, TestModule);

      expect(() => {
        container.getProvider(Nest, TestModule);
      }).toThrow(message);
    });
  });
});
