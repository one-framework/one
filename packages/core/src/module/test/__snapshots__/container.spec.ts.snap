// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneContainer addModule should not add module if already exists in collection 1`] = `
Array [
  Array [
    "586e98de7d09e92da6f61c63d5e0cb3360797533",
    OneModule {
      "container": OneContainer {
        "dynamicModulesMetadata": Map {},
        "globalModules": Set {},
        "moduleCompiler": ModuleCompiler {
          "moduleTokenFactory": ModuleTokenFactory {},
        },
        "moduleOrder": Set {},
        "modules": Map {
          "586e98de7d09e92da6f61c63d5e0cb3360797533" => [Circular],
        },
        "providerTokens": Set {},
      },
      "created": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
      "exports": Set {},
      "factoryValues": Map {},
      "imports": Set {},
      "injector": Injector {
        "_bindingDictionary": Lookup {
          "_map": Map {
            [Function] => Array [
              Binding {
                "activated": false,
                "cache": [Circular],
                "constraint": [Function],
                "dynamicValue": null,
                "factory": null,
                "id": 1,
                "implementationType": null,
                "onActivation": null,
                "provider": null,
                "scope": "Transient",
                "serviceIdentifier": [Function],
                "type": "ConstantValue",
              },
            ],
            [Function] => Array [
              Binding {
                "activated": false,
                "cache": OneContainer {
                  "dynamicModulesMetadata": Map {},
                  "globalModules": Set {},
                  "moduleCompiler": ModuleCompiler {
                    "moduleTokenFactory": ModuleTokenFactory {},
                  },
                  "moduleOrder": Set {},
                  "modules": Map {
                    "586e98de7d09e92da6f61c63d5e0cb3360797533" => [Circular],
                  },
                  "providerTokens": Set {},
                },
                "constraint": [Function],
                "dynamicValue": null,
                "factory": null,
                "id": 2,
                "implementationType": null,
                "onActivation": null,
                "provider": null,
                "scope": "Transient",
                "serviceIdentifier": [Function],
                "type": "ConstantValue",
              },
            ],
            Symbol(InjectionToken<Ref<Module>>) => Array [
              Binding {
                "activated": false,
                "cache": [Circular],
                "constraint": [Function],
                "dynamicValue": null,
                "factory": null,
                "id": 3,
                "implementationType": null,
                "onActivation": null,
                "provider": null,
                "scope": "Transient",
                "serviceIdentifier": Symbol(InjectionToken<Ref<Module>>),
                "type": "ConstantValue",
              },
            ],
          },
        },
        "_metadataReader": MetadataReader {},
        "_middleware": null,
        "_snapshots": Array [],
        "id": 0,
        "options": Object {
          "autoBindInjectable": false,
          "defaultScope": "Transient",
          "skipBaseClassChecks": true,
        },
        "parent": null,
      },
      "lazyInject": [Function],
      "providers": Set {},
      "scope": Array [],
      "target": [Function],
    },
  ],
]
`;
